### PR TITLE
Offboard auth

### PR DESCRIPTION
**Pull request by peterbarker**

### CODE REVIEW

File: libraries/GCS_MAVLink/#GCS.cpp# [hunk 1]
```diff
@@ -0,0 +1,55 @@
+#include "GCS.h"
```
Comment (WickedShell) [inline libraries/GCS_MAVLink/#GCS.cpp#]: What is this file? (the name is really weird)

File: libraries/GCS_MAVLink/GCS.h.orig [hunk 2]
```diff
@@ -0,0 +1,580 @@
+/// @file	GCS.h
```
Comment (WickedShell) [inline libraries/GCS_MAVLink/GCS.h.orig]: Git/patch problems?

Comment (peterbarker) [reply libraries/GCS_MAVLink/GCS.h.orig]: Yes.  Fixed.

File: libraries/GCS_MAVLink/#GCS.cpp# (hunk 1 above)
Comment (peterbarker) [reply libraries/GCS_MAVLink/#GCS.cpp#]: emacs file indicating a file is currently being edited.  Fixed.

**Comment by peterbarker** [comment]
This seems to be implemented as the "standard" implied by `common.xml` indicates.

However, it looks like the standard may have changed during the implementation of this in PX4 Pro.

The enumeration added isn't used.

There's no way to authorize a specific vehicle to arm with a normal command-ack.

So I'm confused.

File: libraries/AP_Arming/AP_Arming.cpp [hunk 3]
```diff
@@ -70,6 +70,22 @@ const AP_Param::GroupInfo AP_Arming::var_info[] = {
...
+    // @DisplayName: Offboard Authorization Interval
+    // @Description: Interval at which to request offboard auth
+    // @User: Advanced
+    // @Units: s
+    // @Increment: 0.1
+    AP_GROUPINFO("OBAUTHINT",     6,     AP_Arming,  offboard_auth_interval,  10),
+
+    // @Param: OBDUTHTMX
+    // @DisplayName: Offboard Authorization Time Max
+    // @Description: Time authorisation is assumed valid for; 0 to disable
```
Comment (magicrub) [inline libraries/AP_Arming/AP_Arming.cpp]: spelling authorisation: -s +z

File: libraries/AP_Arming/AP_Arming.cpp [hunk 4]
```diff
@@ -70,6 +70,22 @@ const AP_Param::GroupInfo AP_Arming::var_info[] = {
     // @User: Standard
     AP_GROUPINFO("VOLT2_MIN",     5,     AP_Arming,  _min_voltage[1],  0),
 
+    // @Param: OBAUTHINT
+    // @DisplayName: Offboard Authorization Interval
```
Comment (magicrub) [inline libraries/AP_Arming/AP_Arming.cpp]: if the param is OB* then lets' have the displayname and desc emphasize the other B like "Off-Board.."

File: libraries/AP_Arming/AP_Arming.cpp [hunk 5]
```diff
@@ -70,6 +70,22 @@ const AP_Param::GroupInfo AP_Arming::var_info[] = {
...
+    // @Description: Interval at which to request offboard auth
+    // @User: Advanced
+    // @Units: s
+    // @Increment: 0.1
+    AP_GROUPINFO("OBAUTHINT",     6,     AP_Arming,  offboard_auth_interval,  10),
```
Comment (magicrub) [inline libraries/AP_Arming/AP_Arming.cpp]: Increment: 0.1 is wrong, Should be 1

File: libraries/AP_Arming/AP_Arming.cpp [hunk 6]
```diff
@@ -658,3 +675,35 @@ bool AP_Arming::rc_checks_copter_sub(const bool display_failure, const RC_Channe
     }
     return ret;
 }
+
+// expected to be called at some rate to do offboard auth as required
+bool AP_Arming::offboard_authorization_ok(bool report)
+{
+    if (offboard_auth_maxt == 0) {
```
Comment (magicrub) [inline libraries/AP_Arming/AP_Arming.cpp]: <= 0

File: libraries/AP_Arming/AP_Arming.cpp (hunk 3 above)
Comment (magicrub) [reply libraries/AP_Arming/AP_Arming.cpp]: "OBDUTHTMX" -> You're missing an A

**Comment by magicrub** [comment]
closes https://github.com/ArduPilot/ardupilot/issues/6405

**Comment by peterbarker** [comment]
@magicrub I've addressed the points you raised.

With the exception of the naming of the variable - the A is missing because we're right on the 16 character limit.  The `ARMING_` prefix doesn't help :-)

**Comment by magicrub** [comment]
I'd say rename "offboard" to "external" and remove the OB in front of the strings. ARMING_AUTH_*

File: libraries/GCS_MAVLink/GCS_Common.cpp.orig [hunk 7]
```diff
@@ -0,0 +1,2567 @@
+/*
```
Comment (WickedShell) [inline libraries/GCS_MAVLink/GCS_Common.cpp.orig]: more emacs files I guess?

**Comment by peterbarker** [comment]
On Thu, 30 Nov 2017, WickedShell wrote:

> more emacs files I guess?

*sigh*.  No, more patch files.  Fixed.

File: libraries/AP_Param/AP_Param.cpp [hunk 8]
```diff
@@ -191,7 +191,10 @@ bool AP_Param::check_group_info(const struct AP_Param::GroupInfo *  group_info,
             return false;
         }
         if (prefix_length + strlen(group_info[i].name) > 16) {
-            Debug("suffix is too long in %s", group_info[i].name);
```
Comment (amilcarlucas) [inline libraries/AP_Param/AP_Param.cpp]: Isn't there a #define somewhere for the 16 ?

**Comment by auturgy** [comment]
Can this be a priority please? If it is functional and safe to merge, LGTM.
We know the next step is an alternate method anyway, so debate over naming and the like should perhaps be deferred.

File: libraries/GCS_MAVLink/GCS_Common.cpp [hunk 9]
```diff
@@ -2059,7 +2086,9 @@ MAV_RESULT GCS_MAVLINK::handle_command_request_autopilot_capabilities(const mavl
 
 MAV_RESULT GCS_MAVLINK::handle_command_do_send_banner(const mavlink_command_long_t &packet)
 {
+    ::fprintf(stderr, "Sending banner\n");
```
Comment (WickedShell) [inline libraries/GCS_MAVLink/GCS_Common.cpp]: Debug?

**Comment by peterbarker** [comment]
This was discussed on the dev call today.

It was discovered that the `COMMAND_ACK` extensions currently present in the mavlink master repository are fundamentally flawed given mavlink2's 0-filling default for extensions (0 is a valid value for target system id / component id).

Since those extensions are required for implementation of this authorization message according to "standard", we simply can not implement this feature using the COMMAND_LONG message.

It was decided that a pair of mavlink2 messages would be created in order to accomplish the same goal.  one message for requesting authorization and a second message granting authorization.

@auturgy We can't merge this after correcting for me not understanding the ack protocol - if we did we'd be locked into mavlink2 extensions on `COMMAND_ACK` which are no good.

**Pull request closed without merging**